src/main/resources/db/migration/V1_1__create_relations.sql

```diff
@@ -0,0 +1,148 @@
+
+CREATE TABLE CategoryEntity (
+    id UUID PRIMARY KEY UNIQUE
+);
+
+CREATE TABLE ProductToCategoryEntity (
+    id UUID PRIMARY KEY UNIQUE DEFAULT uuid_generate_v4(),
+    productId UUID NOT NULL,
+    categoryId UUID NOT NULL,
+    UNIQUE (productId, categoryId)
+);
+
+CREATE TABLE DiscountEntity (
+    id UUID PRIMARY KEY UNIQUE DEFAULT uuid_generate_v4(),
+    discount DOUBLE PRECISION NOT NULL,
+    maxUsagesPerUser INTEGER NULL,
+    validFrom TIMESTAMPTZ NOT NULL,
+    validUntil TIMESTAMPTZ NOT NULL,
+    minOrderAmount INTEGER NULL
+);
+
+CREATE TABLE CouponEntity (
+    id UUID PRIMARY KEY UNIQUE DEFAULT uuid_generate_v4(),
+    usages INTEGER NOT NULL DEFAULT 0,
+    maxUsages INTEGER NULL,
+    validFrom TIMESTAMPTZ NOT NULL,
+    validUntil TIMESTAMPTZ NOT NULL,
+    code VARCHAR(255) NOT NULL,
+    discountId UUID NOT NULL,
+    FOREIGN KEY (discountId) REFERENCES DiscountEntity(id),
+    UNIQUE (code)
+);
+
+CREATE TABLE CouponToUserEntity (
+    id UUID PRIMARY KEY UNIQUE DEFAULT uuid_generate_v4(),
+    couponId UUID NOT NULL,
+    userId UUID NOT NULL,
+    FOREIGN KEY (couponId) REFERENCES CouponEntity(id),
+    FOREIGN KEY (userId) REFERENCES UserEntity(id),
+    UNIQUE (couponId, userId)
+);
+
+CREATE TABLE DiscountToProductEntity (
+    id UUID PRIMARY KEY UNIQUE DEFAULT uuid_generate_v4(),
+    discountId UUID NOT NULL,
+    productId UUID NOT NULL,
+    FOREIGN KEY (discountId) REFERENCES DiscountEntity(id),
+    FOREIGN KEY (productId) REFERENCES ProductEntity(id),
+    UNIQUE (discountId, productId)
+);
+
+CREATE TABLE DiscountToCategoryEntity (
+    id UUID PRIMARY KEY UNIQUE DEFAULT uuid_generate_v4(),
+    discountId UUID NOT NULL,
+    categoryId UUID NOT NULL,
+    FOREIGN KEY (discountId) REFERENCES DiscountEntity(id),
+    FOREIGN KEY (categoryId) REFERENCES CategoryEntity(id),
+    UNIQUE (discountId, categoryId)
+);
+
+CREATE TABLE DiscountToProductVariantEntity (
+    id UUID PRIMARY KEY UNIQUE DEFAULT uuid_generate_v4(),
+    discountId UUID NOT NULL,
+    productVariantId UUID NOT NULL,
+    FOREIGN KEY (discountId) REFERENCES DiscountEntity(id),
+    FOREIGN KEY (productVariantId) REFERENCES ProductVariantEntity(id),
+    UNIQUE (discountId, productVariantId)
+);
+
+CREATE TABLE DiscountUsageEntity(
+    id UUID PRIMARY KEY UNIQUE DEFAULT uuid_generate_v4(),
+    discountId UUID NOT NULL,
+    userId UUID NOT NULL,
+    usages BIGINT NOT NULL DEFAULT 0,
+    FOREIGN KEY (discountId) REFERENCES DiscountEntity(id),
+    FOREIGN KEY (userId) REFERENCES UserEntity(id),
+    UNIQUE (discountId, userId)
+);
+)
+
+CREATE OR REPLACE FUNCTION update_coupon_usages()
+RETURNS TRIGGER AS $$
+DECLARE
+    current_usages INTEGER;
+    max_usages INTEGER;
+BEGIN
+    SELECT couponEntity.usages, couponEntity.maxUsages
+    INTO current_usages, max_usages
+    FROM CouponEntity couponEntity
+    WHERE couponEntity.id = NEW.couponId;
+
+    current_usages := current_usages + 1;
+
+    IF current_usages > max_usages THEN
```

Comment:
Does this work correctly when `max_usages == NULL`?
```suggestion
    IF max_usages IS NOT NULL AND current_usages > max_usages THEN
```